stable/database/files/nuobackup

```diff
@@ -42,6 +44,64 @@ labels=""
 semaphore=""
 timeout=1800
 backup_root=$BACKUP_DIR
+wait_for_num_processes=0
+wait_timeout=300
+
+function await() {
+   local timeout=$1
+   shift
+   local end_time=$(( $(date -u +%s) + $timeout ))
+   local rc=0
+   while [ "$(date -u +%s)" -lt "$end_time" ]; do
+      eval "$*"
+      rc=$?
+      if [ $rc -eq 0 ]; then
+         return $rc
+      fi
+      sleep 1
+   done
+   return 124
+}
+
+function check_backup_processes() {
+   declare -a label_patterns=()
+   local all_process_labels=""
+
+   for ((i=0; i<${#labels_arr[@]}; i+=2)); do
+     label_patterns[${#label_patterns[@]}]="\"${labels_arr[i]}\"\s*:\s*\"${labels_arr[i+1]}\""
+   done
+
+   all_process_labels="$(nuocmd --api-server $NUOCMD_API_SERVER show database \
+      --db-name $db_name \
+      --process-format '#{engine_type} {state} {labels}')" || return $?
+
+   num_processes=0
+   while read process_labels; do
```

Comment:
The label patterns are specified via standard input? Also, the labels are printed in the following format:
```
{'label': 'value'[, ...]}
```

Are we expecting users to specify something like `'type': 'backup'`, for example?

Reply:
Not sure I understand the question completely. The expected labels are formed from the value of `--group` + values of `--labels` options. Currently `nuobackup --labels ...` is not used anywhere so effectively we are searching for SMs that have the following label `'backup': 'cluster0'`.
There is an explicit label `backup` injected in HC SMs which value is the backup group:
```
        args:
          - "nuosm"
          ...
          - "--labels"
          - "backup {{ include "hotcopy.group" . }} {{- include "opt.key-values" .Values.database.sm.labels }}"
```

Reply:
Unless I'm misunderstanding some Bashism, `read process_labels` reads from standard input into the `process_labels` variable. If standard input is not a tty, then this loop is skipped entirely. At least that's what I think the code does from reading it, but I haven't tested it.

Have you done any manual testing that actually causes us to enter the loop? If this is not used anywhere, I would just remove the label-matching functionality.

Reply:
Yes, manual testing is done and this is how we await for the processes to become running so that we make the initial backup.

See line 97 --- the `stdin` is replaced with a specific string so that `read`. For example:
```
while read file; do
  echo $file
  # Do something
done <<< "$(ls -1 /tmp)"

Reply:
Ah, I didn't even get that far... what a wonderful Bash idiom.

Thanks for the explanation.